twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java

```diff
@@ -220,5 +250,31 @@ public ContainerLiveNodeData getLiveNodeData() {
     public void kill() {
       processController.cancel();
     }
+
+    private void killAndWait(int maxWaitSecs) {
+      Stopwatch watch = new Stopwatch();
+      watch.start();
+      int tries = 0;
+      while (watch.elapsedTime(TimeUnit.SECONDS) < maxWaitSecs) {
+        // Kill the application
+        try {
+          ++tries;
+          kill();
+        } catch (Exception e) {
+          LOG.error("Exception while killing runnable {}, instance {}", runnable, instanceId, e);
+        }
+
+        // Wait on the shutdownLatch,
+        // if the runnable has stopped then the latch will be count down by completed() method
+        if (Uninterruptibles.awaitUninterruptibly(shutdownLatch, 10, TimeUnit.SECONDS)) {
+          // Runnable has stopped now
+          return;
+        }
+      }
+
+      // Timeout reached, runnable has not stopped
+      LOG.error("Failed to kill runnable {}, instance {} after {} tries", runnable, instanceId, tries);
```

Comment:
Showing the number of tries is quite artificial since the retry is based on time. I think it's better to just say failed to kill after n seconds.